pkg/app/piped/executor/lambda/rollback.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func (e *rollbackExecutor) ensureRollback(ctx context.Context) model.StageStatus {
+	// Not rollback in case this is the first deployment.
+	if e.Deployment.RunningCommitHash == "" {
+		e.LogPersister.Errorf("Unable to determine the last deployed commit to rollback. It seems this is the first deployment.")
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	runningDS, err := e.RunningDSP.GetReadOnly(ctx, e.LogPersister)
+	if err != nil {
+		e.LogPersister.Errorf("Failed to prepare running deploy source data (%v)", err)
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	deployCfg := runningDS.DeploymentConfig.LambdaDeploymentSpec
+	if deployCfg == nil {
+		e.LogPersister.Errorf("Malformed deployment configuration: missing LambdaDeploymentSpec")
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	cloudProviderName, cloudProviderCfg, found := findCloudProvider(&e.Input)
+	if !found {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	fm, ok := loadFunctionManifest(&e.Input, deployCfg.Input.FunctionManifestFile, runningDS)
+	if !ok {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	if !rollback(ctx, &e.Input, cloudProviderName, cloudProviderCfg, fm) {
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	return model.StageStatus_STAGE_SUCCESS
+}
+
+func rollback(ctx context.Context, in *executor.Input, cloudProviderName string, cloudProviderCfg *config.CloudProviderLambdaConfig, fm provider.FunctionManifest) bool {
+	in.LogPersister.Infof("Start rollback the lambda function: %s to original stage", fm.Spec.Name)
+	client, err := provider.DefaultRegistry().Client(cloudProviderName, cloudProviderCfg, in.Logger)
+	if err != nil {
+		in.LogPersister.Errorf("Unable to create Lambda client for the provider %s: %v", cloudProviderName, err)
+		return false
+	}
+
+	// Restore original traffic config from metadata store.
+	originalTrafficKeyName := fmt.Sprintf("%s-%s-original", fm.Spec.Name, in.Deployment.RunningCommitHash)
+	originalTrafficCfgData, ok := in.MetadataStore.Get(originalTrafficKeyName)
+	if !ok {
+		in.LogPersister.Errorf("Unable to prepare original traffic config to rollback Lambda function %s. It seems this is the first deployment.", fm.Spec.Name)
+		return false
+	}
+
+	originalTrafficCfg := provider.RoutingTrafficConfig{}
+	if err := json.Unmarshal([]byte(originalTrafficCfgData), &originalTrafficCfg); err != nil {
+		in.LogPersister.Errorf("Unable to prepare original traffic config to rollback Lambda function %s: %v", fm.Spec.Name, err)
+		return false
+	}
+
+	// Restore promoted traffic config from metadata store.
+	promotedTrafficKeyName := fmt.Sprintf("%s-%s-promote", fm.Spec.Name, in.Deployment.RunningCommitHash)
+	promotedTrafficCfgData, ok := in.MetadataStore.Get(promotedTrafficKeyName)
+	// If there is no previous promoted traffic config, which mean no promote run previously so no need to do anything to rollback.
+	if !ok {
+		in.LogPersister.Info("No promoted traffic config found. No need to rollback current remote traffic config.")
```

Comment:
nit: "It seems the traffic has not been changed during the deployment process. No need to rollback the traffic config."